In [4]:
# First I need the proper libraries
import requests
from bs4 import BeautifulSoup
import pandas as pd

# This imports my personal API key
import api

In [ ]:
# Covid Act Now Guide:
# https://api.covidactnow.org/v2/country/US.json?apiKey=YOUR_KEY_HERE
# https://api.covidactnow.org/v2/country/US.timeseries.json?apiKey=YOUR_KEY_HERE
# https://api.covidactnow.org/v2/country/US.timeseries.csv?apiKey=YOUR_KEY_HERE

In [11]:
response = requests.get(f'https://api.covidactnow.org/v2/country/US.json?apiKey={api.key}')
html = response.text
html

'{"fips": "0", "country": "US", "state": null, "county": null, "level": "country", "lat": null, "locationId": "iso1:us", "long": null, "population": 331486822, "metrics": {"testPositivityRatio": 0.025, "testPositivityRatioDetails": {"source": "other"}, "caseDensity": 9.2, "contactTracerCapacityRatio": null, "infectionRate": 0.79, "infectionRateCI90": 0.1, "icuCapacityRatio": 0.73, "vaccinationsInitiatedRatio": 0.769, "vaccinationsCompletedRatio": 0.654, "vaccinationsAdditionalDoseRatio": 0.291}, "riskLevels": {"overall": 1, "testPositivityRatio": 0, "caseDensity": 1, "contactTracerCapacityRatio": 4, "infectionRate": 0, "icuCapacityRatio": 1}, "cdcTransmissionLevel": 2, "actuals": {"cases": 79532295, "deaths": 967455, "positiveTests": 75868952, "negativeTests": 735428084, "contactTracers": 65999, "hospitalBeds": {"capacity": 751837, "currentUsageTotal": 571394, "currentUsageCovid": 22590}, "icuBeds": {"capacity": 77818, "currentUsageTotal": 56648, "currentUsageCovid": 3391}, "newCases":

"fips": "0"
"country": "US", 
"state": null, 
"county": null, 
"level": "country", 
"lat": null, 
"locationId": "iso1:us", 
"long": null, 
"population": 331486822, 
"metrics": {"testPositivityRatio": 0.025, "testPositivityRatioDetails": {"source": "other"}, "caseDensity": 9.2, "contactTracerCapacityRatio": null, "infectionRate": 0.79, "infectionRateCI90": 0.1, "icuCapacityRatio": 0.73, "vaccinationsInitiatedRatio": 0.769, "vaccinationsCompletedRatio": 0.654, "vaccinationsAdditionalDoseRatio": 0.291},

"riskLevels": {"overall": 1, "testPositivityRatio": 0, "caseDensity": 1, "contactTracerCapacityRatio": 4, "infectionRate": 0, "icuCapacityRatio": 1}, 
"cdcTransmissionLevel": 2, 

"actuals": {"cases": 79532295, "deaths": 967455, "positiveTests": 75868952, "negativeTests": 735428084, "contactTracers": 65999, "hospitalBeds": {"capacity": 751837, "currentUsageTotal": 571394, "currentUsageCovid": 22590}, "icuBeds": {"capacity": 77818, "currentUsageTotal": 56648, "currentUsageCovid": 3391}, "newCases": 39384, "newDeaths": 2017, "vaccinesDistributed": 697440555, "vaccinationsInitiated": 254823742, "vaccinationsCompleted": 216897061, "vaccinationsAdditionalDose": 96335336, "vaccinesAdministered": 557902891, "vaccinesAdministeredDemographics": null, "vaccinationsInitiatedDemographics": null}, 

"annotations": {"cases": {"sources": [{"type": "NYTimes", "url": "https://github.com/nytimes/covid-19-data", "name": "The New York Times"}], "anomalies": []}, "deaths": {"sources": [{"type": "NYTimes", "url": "https://github.com/nytimes/covid-19-data", "name": "The New York Times"}], "anomalies": []}, "positiveTests": null, "negativeTests": null, "contactTracers": null, "hospitalBeds": null, "icuBeds": null, "newCases": null, "newDeaths": {"sources": [], "anomalies": [{"date": "2020-06-25", "type": "zscore_outlier", "original_observation": 2466.0}]}, "vaccinesDistributed": {"sources": [{"type": "other", "url": "https://data.cdc.gov/Vaccinations/COVID-19-Vaccinations-in-the-United-States-Jurisdi/unsk-b7fc", "name": "Centers for Disease Control and Prevention"}], "anomalies": []}, "vaccinationsInitiated": {"sources": [{"type": "other", "url": "https://data.cdc.gov/Vaccinations/COVID-19-Vaccinations-in-the-United-States-Jurisdi/unsk-b7fc", "name": "Centers for Disease Control and Prevention"}], "anomalies": []}, "vaccinationsCompleted": {"sources": [{"type": "other", "url": "https://data.cdc.gov/Vaccinations/COVID-19-Vaccinations-in-the-United-States-Jurisdi/unsk-b7fc", "name": "Centers for Disease Control and Prevention"}], "anomalies": []}, "vaccinationsAdditionalDose": {"sources": [{"type": "other", "url": "https://data.cdc.gov/Vaccinations/COVID-19-Vaccinations-in-the-United-States-Jurisdi/unsk-b7fc", "name": "Centers for Disease Control and Prevention"}], "anomalies": []}, "vaccinesAdministered": {"sources": [{"type": "other", "url": "https://data.cdc.gov/Vaccinations/COVID-19-Vaccinations-in-the-United-States-Jurisdi/unsk-b7fc", "name": "Centers for Disease Control and Prevention"}], "anomalies": []}, "testPositivityRatio": null, "caseDensity": {"sources": [{"type": "NYTimes", "url": "https://github.com/nytimes/covid-19-data", "name": "The New York Times"}], "anomalies": []}, "contactTracerCapacityRatio": null, "infectionRate": {"sources": [{"type": "NYTimes", "url": "https://github.com/nytimes/covid-19-data", "name": "The New York Times"}], "anomalies": []}, "infectionRateCI90": {"sources": [{"type": "NYTimes", "url": "https://github.com/nytimes/covid-19-data", "name": "The New York Times"}], "anomalies": []}, "icuCapacityRatio": null, "vaccinationsInitiatedRatio": null, "vaccinationsCompletedRatio": null, "vaccinationsAdditionalDoseRatio": null}, "lastUpdatedDate": "2022-03-18", "url": null}'

In [10]:
# Create a csv of the scraped data
html.to_csv('aggregated_us_data.csv')

AttributeError: 'str' object has no attribute 'to_csv'

In [12]:
# Make a dataframe out of it
df = pd.DataFrame(html)
df

ValueError: DataFrame constructor not properly called!

In [ ]:
# initialize list of lists
data = [['tom', 10], ['nick', 15], ['juli', 14]]
 
# Create the pandas DataFrame
df = pd.DataFrame(data, columns = ['Name', 'Age'])
 

In [14]:
import requests
import csv

# Define the remote URL
# url = "https://query1.finance.yahoo.com/v7/finance/download/SPY"

# Send HTTP GET request via requests
# data = requests.get(url)
data = requests.get(f'https://api.covidactnow.org/v2/country/US.timeseries.csv?apiKey={api.key}')

# Convert to iterator by splitting on \n chars
lines = data.text.splitlines()

# Parse as CSV object
reader = csv.reader(lines)

# View Result
for row in reader:
    print(row)


['date', 'country', 'state', 'county', 'fips', 'lat', 'long', 'locationId', 'actuals.cases', 'actuals.deaths', 'actuals.positiveTests', 'actuals.negativeTests', 'actuals.contactTracers', 'actuals.hospitalBeds.capacity', 'actuals.hospitalBeds.currentUsageTotal', 'actuals.hospitalBeds.currentUsageCovid', 'unused1', 'actuals.icuBeds.capacity', 'actuals.icuBeds.currentUsageTotal', 'actuals.icuBeds.currentUsageCovid', 'unused2', 'actuals.newCases', 'actuals.vaccinesDistributed', 'actuals.vaccinationsInitiated', 'actuals.vaccinationsCompleted', 'metrics.testPositivityRatio', 'metrics.testPositivityRatioDetails', 'metrics.caseDensity', 'metrics.contactTracerCapacityRatio', 'metrics.infectionRate', 'metrics.infectionRateCI90', 'unused3', 'unused4', 'metrics.icuCapacityRatio', 'riskLevels.overall', 'metrics.vaccinationsInitiatedRatio', 'metrics.vaccinationsCompletedRatio', 'actuals.newDeaths', 'actuals.vaccinesAdministered', 'riskLevels.caseDensity', 'cdcTransmissionLevel', 'actuals.vaccination

In [17]:
# Create the pandas DataFrame
df = pd.DataFrame(data= lines, columns= lines[0])
df

TypeError: Index(...) must be called with a collection of some kind, 'date,country,state,county,fips,lat,long,locationId,actuals.cases,actuals.deaths,actuals.positiveTests,actuals.negativeTests,actuals.contactTracers,actuals.hospitalBeds.capacity,actuals.hospitalBeds.currentUsageTotal,actuals.hospitalBeds.currentUsageCovid,unused1,actuals.icuBeds.capacity,actuals.icuBeds.currentUsageTotal,actuals.icuBeds.currentUsageCovid,unused2,actuals.newCases,actuals.vaccinesDistributed,actuals.vaccinationsInitiated,actuals.vaccinationsCompleted,metrics.testPositivityRatio,metrics.testPositivityRatioDetails,metrics.caseDensity,metrics.contactTracerCapacityRatio,metrics.infectionRate,metrics.infectionRateCI90,unused3,unused4,metrics.icuCapacityRatio,riskLevels.overall,metrics.vaccinationsInitiatedRatio,metrics.vaccinationsCompletedRatio,actuals.newDeaths,actuals.vaccinesAdministered,riskLevels.caseDensity,cdcTransmissionLevel,actuals.vaccinationsAdditionalDose,metrics.vaccinationsAdditionalDoseRatio' was passed

In [23]:
# The first element has a list of the column names. I want to make those into a variable
lines[0].replace(',' , '","')

'date","country","state","county","fips","lat","long","locationId","actuals.cases","actuals.deaths","actuals.positiveTests","actuals.negativeTests","actuals.contactTracers","actuals.hospitalBeds.capacity","actuals.hospitalBeds.currentUsageTotal","actuals.hospitalBeds.currentUsageCovid","unused1","actuals.icuBeds.capacity","actuals.icuBeds.currentUsageTotal","actuals.icuBeds.currentUsageCovid","unused2","actuals.newCases","actuals.vaccinesDistributed","actuals.vaccinationsInitiated","actuals.vaccinationsCompleted","metrics.testPositivityRatio","metrics.testPositivityRatioDetails","metrics.caseDensity","metrics.contactTracerCapacityRatio","metrics.infectionRate","metrics.infectionRateCI90","unused3","unused4","metrics.icuCapacityRatio","riskLevels.overall","metrics.vaccinationsInitiatedRatio","metrics.vaccinationsCompletedRatio","actuals.newDeaths","actuals.vaccinesAdministered","riskLevels.caseDensity","cdcTransmissionLevel","actuals.vaccinationsAdditionalDose","metrics.vaccinationsAddi

In [26]:
# Okay lets do this the long way
col_list= ["date",
"country",
"state",
"county",
"fips",
"lat",
"long",
"locationId",
"actuals.cases",
"actuals.deaths",
"actuals.positiveTests",
"actuals.negativeTests",
"actuals.contactTracers",
"actuals.hospitalBeds.capacity",
"actuals.hospitalBeds.currentUsageTotal",
"actuals.hospitalBeds.currentUsageCovid",
"unused1",
"actuals.icuBeds.capacity",
"actuals.icuBeds.currentUsageTotal",
"actuals.icuBeds.currentUsageCovid",
"unused2","actuals.newCases",
"actuals.vaccinesDistributed",
"actuals.vaccinationsInitiated",
"actuals.vaccinationsCompleted",
"metrics.testPositivityRatio",
"metrics.testPositivityRatioDetails",
"metrics.caseDensity",
"metrics.contactTracerCapacityRatio",
"metrics.infectionRate",
"metrics.infectionRateCI90",
"unused3",
"unused4",
"metrics.icuCapacityRatio",
"riskLevels.overall",
"metrics.vaccinationsInitiatedRatio",
"metrics.vaccinationsCompletedRatio",
"actuals.newDeaths",
"actuals.vaccinesAdministered",
"riskLevels.caseDensity",
"cdcTransmissionLevel",
"actuals.vaccinationsAdditionalDose",
"metrics.vaccinationsAdditionalDoseRatio"]

In [ ]:
# Send HTTP GET request via requests
# data = requests.get(url)
data = requests.get(f'https://api.covidactnow.org/v2/country/US.timeseries.csv?apiKey={api.key}')

# Convert to iterator by splitting on \n chars
lines = data.text.splitlines()

# Parse as CSV object
reader = csv.reader(lines)

# View Result
for row in reader:
    print(row)

In [29]:
# Collecting the data from the API and converting it into a csv
csv_url = f'https://api.covidactnow.org/v2/country/US.timeseries.csv?apiKey={api.key}'
req = requests.get(csv_url)
url_content = req.content
csv_file = open('aggregated_us_data.csv', 'wb')

csv_file.write(url_content)
csv_file.close()

In [30]:
# Read the csv into a dataframe
covid = pd.read_csv('aggregated_us_data.csv', index_col=0)
covid.head()

country  state  county  fips  lat  long locationId  actuals.cases  \
date                                                                           
2020-01-21      US    NaN     NaN     0  NaN   NaN    iso1:us            1.0   
2020-01-22      US    NaN     NaN     0  NaN   NaN    iso1:us            1.0   
2020-01-23      US    NaN     NaN     0  NaN   NaN    iso1:us            1.0   
2020-01-24      US    NaN     NaN     0  NaN   NaN    iso1:us            2.0   
2020-01-25      US    NaN     NaN     0  NaN   NaN    iso1:us            3.0   

            actuals.deaths  actuals.positiveTests  ...  \
date                                               ...   
2020-01-21             0.0                    NaN  ...   
2020-01-22             0.0                    NaN  ...   
2020-01-23             0.0                    NaN  ...   
2020-01-24             0.0                    NaN  ...   
2020-01-25             0.0                    NaN  ...   

            metrics.icuCapacityRatio  riskLevels.overall  \
date                                                       
2020-01-21                       NaN                   0   
2020-01-22                       NaN                   0   
2020-01-23                       NaN                   0   
2020-01-24                       NaN                   0   
2020-01-25                       NaN                   0   

            metrics.vaccinationsInitiatedRatio  \
date                                             
2020-01-21                                 NaN   
2020-01-22                                 NaN   
2020-01-23                                 NaN   
2020-01-24                                 NaN   
2020-01-25                                 NaN   

            metrics.vaccinationsCompletedRatio  actuals.newDeaths  \
date                                                                
2020-01-21                                 NaN                NaN   
2020-01-22                                 NaN                0.0   
2020-01-23                                 NaN                0.0   
2020-01-24                                 NaN                0.0   
2020-01-25                                 NaN                0.0   

            actuals.vaccinesAdministered  riskLevels.caseDensity  \
date                                                               
2020-01-21                           NaN                       0   
2020-01-22                           NaN                       0   
2020-01-23                           NaN                       0   
2020-01-24                           NaN                       0   
2020-01-25                           NaN                       0   

            cdcTransmissionLevel  actuals.vaccinationsAdditionalDose  \
date                                                                   
2020-01-21                     0                                 NaN   
2020-01-22                     0                                 NaN   
2020-01-23                     0                                 NaN   
2020-01-24                     0                                 NaN   
2020-01-25                     0                                 NaN   

            metrics.vaccinationsAdditionalDoseRatio  
date                                                 
2020-01-21                                      NaN  
2020-01-22                                      NaN  
2020-01-23                                      NaN  
2020-01-24                                      NaN  
2020-01-25                                      NaN  

[5 rows x 42 columns]

In [ ]:
# Now i need a function to feed into the dataframe
def parse_data(lines):
    headline = article.h2.text
    date = article.p.text.strip()
    author = article.select('.text-right')[0].text.strip()[3:]
    content = article.select('p')[-1].text
   
    return {
        'headline': headline, 'date': date, 'author': author,
        'content': content
    }


In [28]:
# Create the pandas DataFrame
df = pd.DataFrame(data= data, columns= col_list)
df

ValueError: Shape of passed values is (1093, 1), indices imply (1093, 43)